accepted/future-releases/inline-classes/feature-specification.md

```diff
@@ -0,0 +1,969 @@
+  final Object it;
+
+  TinyJson(this.it);
+
+  Iterable<num> get leaves sync* {
+    var localIt = it; // To get promotion.
+    if (localIt is num) {
+      yield localIt;
+    } else if (localIt is List<dynamic>) {
+      for (var element in localIt) {
+        yield* TinyJson(element).leaves;
+      }
+    } else {
+      throw "Unexpected object encountered in TinyJson value";
+    }
+  }
+}
+
+void main() {
+  var tiny = TinyJson(<dynamic>[<dynamic>[1, 2], 3, <dynamic>[]]);
+  print(tiny.leaves);
+  tiny.add("Hello!"); // Error.
+}
+```
+
+This is similar to the inline type in that it enforces the use of
+specific operations only (here we have just one: `leaves`), and it
+makes it an error to use instance methods of the representation (e.g.,
+`add`).
+
+Creation of wrapper objects consumes space and time. In the case where
+we wish to work on an entire data structure, we'd need to wrap each
+object as we navigate the data structure. For instance, we need to
+create a wrapper `TinyJson(element)` in order to invoke `leaves`
+recursively.
+
+In contrast, the inline class is zero-cost, in the sense that it does
+_not_ use a wrapper object, it enforces the desired discipline
+statically. In the inline class, the invocation of `TinyJson(element)`
+in the body of `leaves` can be eliminated entirely by inlining.
+
+Inline classes are static in nature, like extension members: an inline
+class declaration may declare some type parameters. The type
+parameters will be bound to types which are determined by the static
+type of the receiver. Similarly, members of an inline type are resolved
+statically, i.e., if `tiny.leaves` is an invocation of an inline getter
+`leaves`, then the declaration named `leaves` whose body is executed
+is determined at compile-time. There is no support for late binding of
+an inline member, and hence there is no notion of overriding. In return
+for this lack of expressive power, we get improved performance.
+
+
+## Syntax
+
+A rule for `<inlineClassDeclaration>` is added to the grammar, along
+with some rules for elements used in inline class declarations:
+
+```ebnf
+<inlineClassDeclaration> ::=
+  'inline' 'class' <typeIdentifier> <typeParameters>? <interfaces>?
+  '{'
+    (<metadata> <inlineMemberDeclaration>)*
+  '}'
+
+<inlineMemberDeclaration> ::= <classMemberDefinition>
+```
+
+*The token `inline` is not made a built-in identifier: the reserved
+word `class` that occurs right after `inline` serves to disambiguate
+the inline class declaration with a fixed lookahead.*
+
+A few errors can be detected immediately from the syntax:
+
+A compile-time error occurs if the inline class does not declare any
+instance variables, and if it declares two or more instance
+variables. Let `v` be the name of unique instance variable that it
+declares. The declaration of `v` must have the modifier `final`, and it
+can not have the modifier `late`; otherwise a compile-time error
+occurs.
```

Comment:
May it have an initializer?

Reply:
Asked Erik about that in person. Answer was "yes", because it's not a problem to do so. Field initializers are evaluated as part of constructor invocation (the `new`/`const` operators).
The initializer can be non-trivial/const, like `= DateTime.now()` or `++_counter`, so we don't necessarily get the same value each time. Doing `final field = expr;` is really not significantly different than `final field; Foo() : field = expr;`, which we will allow.

Constructors can also be `const`, there is also no problem with that. 

(In practice I'll probably just have something like `const Foo._(this._x);` as a constructor on all my inline classes, and forward all other constructors to that, but you can choose other approaches, just as you can for classes.)

Reply:
Very good! I'll conclude that there is no need to prohibit these initializers. I haven't added further text about this: The inline class static analysis already relies on the analysis of regular classes in many ways (e.g., how to type check the initializer list, the fact that a constructor can be `const` and so on).